### PR TITLE
Use CSI proxy v1 client library

### PR DESCRIPTION
**Pull request by mauriciopoppe**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Adds compatibility with the following versions of the csi-proxy client: disk v1, volume v1, filesystem v1, the plan to upgrade is:

- deploy PD CSI with the csi-proxy v1 (and others) client, keep the beta csi-proxy server on the node, the v1 client would fallback to talking to the beta csi-proxy server
- bump csi-proxy server to v1 on the node, to do this we need to drain the node and restart PD CSI which will then talk to v1 automatically

## Summary of changes:

- add new hostPath mounts to the windows nodes, these mounts are the v1 named pipes, these named pipes don't exist however the implementation is able to understand this fact and it fallbacks to the v1beta2/v1beta1 named pipes
- `safe-mounter_windows.go` was cloned into a v1 and a v1beta version, the differences are the imports (I could also change this to multiple if statements but a clone makes the code easy to follow)
- `NewSafeMounter` creates either a v1 mounter or a v1beta mounter, the `Interface` is casted to the required type on runtime

## Logs of the new implementation:

Test steps:

- create a kubernetes GCE cluster with 2 windows nodes
- choose one of the windows nodes and deploy `csi-proxy.exe`, verify that the v1 named pipes are available

```
PS C:\Users\mauriciopoppe\go\src\github.com\kubernetes-csi\csi-proxy> [System.IO.Directory]::GetFiles("\\.\\pipe\\")
\\.\\pipe\\InitShutdown
\\.\\pipe\\lsass
\\.\\pipe\\ntsvcs
\\.\\pipe\\scerpc
\\.\\pipe\\Winsock2\CatalogChangeListener-364-0
\\.\\pipe\\epmapper
\\.\\pipe\\Winsock2\CatalogChangeListener-1f4-0
\\.\\pipe\\LSM_API_service
\\.\\pipe\\W32TIME_ALT
\\.\\pipe\\eventlog
\\.\\pipe\\Winsock2\CatalogChangeListener-138-0
\\.\\pipe\\wkssvc
\\.\\pipe\\atsvc
\\.\\pipe\\Winsock2\CatalogChangeListener-500-0
\\.\\pipe\\Winsock2\CatalogChangeListener-278-0
\\.\\pipe\\openssh-ssh-agent
\\.\\pipe\\TermSrv_API_service
\\.\\pipe\\Ctx_WinStation_API_service
\\.\\pipe\\SessEnvPublicRpc
\\.\\pipe\\Winsock2\CatalogChangeListener-a7c-0
\\.\\pipe\\srvsvc
\\.\\pipe\\Winsock2\CatalogChangeListener-264-0
\\.\\pipe\\PSHost.132685607009879421.2264.DefaultAppDomain.powershell
\\.\\pipe\\TSVCPIPE-6969a09d-6df7-43e9-8b4d-07ca51b17aeb
\\.\\pipe\\docker_engine
\\.\\pipe\\PSHost.132685607145678539.4928.DefaultAppDomain.powershell
\\.\\pipe\\PIPE_EVENTROOT\CIMV2SCM EVENT PROVIDER
\\.\\pipe\\dockershim
\\.\\pipe\\PSHost.132685608544452389.4896.DefaultAppDomain.powershell
\\.\\pipe\\csi-proxy-filesystem-v1alpha1
\\.\\pipe\\csi-proxy-filesystem-v1beta1
\\.\\pipe\\csi-proxy-filesystem-v1beta2
\\.\\pipe\\csi-proxy-filesystem-v1
\\.\\pipe\\csi-proxy-disk-v1alpha1
\\.\\pipe\\csi-proxy-disk-v1beta1
\\.\\pipe\\csi-proxy-disk-v1beta2
\\.\\pipe\\csi-proxy-disk-v1beta3
\\.\\pipe\\csi-proxy-disk-v1
\\.\\pipe\\csi-proxy-volume-v1alpha1
\\.\\pipe\\csi-proxy-volume-v1beta1
\\.\\pipe\\csi-proxy-volume-v1beta2
\\.\\pipe\\csi-proxy-volume-v1beta3
\\.\\pipe\\csi-proxy-volume-v1
\\.\\pipe\\csi-proxy-smb-v1alpha1
\\.\\pipe\\csi-proxy-smb-v1beta1
\\.\\pipe\\csi-proxy-smb-v1beta2
\\.\\pipe\\csi-proxy-smb-v1
\\.\\pipe\\csi-proxy-system-v1alpha1
\\.\\pipe\\csi-proxy-iscsi-v1alpha1
\\.\\pipe\\csi-proxy-iscsi-v1alpha2
```

- compile PD CSI driver and deploy it the cluster, check the logs of PD CSI driver in both windows nodes

```
# windows node without v1 pipes enabled
I0619 08:14:54.770087    1504 main.go:73] Driver vendor version latest
I0619 08:14:54.843225    1504 safe-mounter_windows.go:35] failed to connect to csi-proxy v1 with error=open \\.\\pipe\\csi-proxy-filesystem-v1: The system cannot find the file specified., will try w
I0619 08:14:54.843225    1504 safe-mounter_windows.go:39] using CSIProxyMounterV1Beta, API Versions Disk: v1beta2, Filesystem: v1beta1, Volume: v1beta1

# windows node with v1 pipes enabled
I0619 08:14:54.500483    3260 main.go:73] Driver vendor version latest
I0619 08:14:54.503411    3260 safe-mounter_windows.go:28] using CSIProxyMounterV1, API Versions Disk: v1, Filesystem: v1, Volume: v1
```

- Test on v1beta: Untaint the v1beta node and taint the v1 node and run the `External Storage` suite from k/k (compatibility with the existing csi-server)

```
GINKGO_PARALLEL_NODES=4 ./hack/ginkgo-e2e.sh --ginkgo.focus="External.*Storage" --ginkgo.skip="\[Disruptive\]|\[Serial\]|\[LinuxOnly\]" --allowed-not-ready-nodes=10 --storage.testdriver="/usr/local/google/home/mauriciopoppe/go/src/sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/test/k8s-integration/config/test-config-windows.yaml" --node-os-distro=windows

Ran 36 of 6637 Specs in 2833.752 seconds
SUCCESS! -- 36 Passed | 0 Failed | 0 Pending | 6601 Skipped
```

- Test on v1: Untaint the v1 node and taint v1beta node and run the `External Storage` suite from k/k

```
GINKGO_PARALLEL_NODES=4 ./hack/ginkgo-e2e.sh --ginkgo.focus="External.*Storage" --ginkgo.skip="\[Disruptive\]|\[Serial\]|\[LinuxOnly\]" --allowed-not-ready-nodes=10 --storage.testdriver="/usr/local/google/home/mauriciopoppe/go/src/sigs.k8s.io/gcp-compute-persistent-disk-csi-driver/test/k8s-integration/config/test-config-windows.yaml" --node-os-distro=windows

Ran 36 of 6637 Specs in 2970.219 seconds
SUCCESS! -- 36 Passed | 0 Failed | 0 Pending | 6601 Skipped
```

- (pending) Run the optional prow job e2e `External Storage` testsuite setting csi-proxy v1 in k/k https://github.com/kubernetes/test-infra/pull/22636

### Next changes:

- bump csi-proxy from v1-rc to v1

**Does this PR introduce a user-facing change?**:

```release-note
Bumped csi-proxy client library to v1.0.0
```

/assign @jingxu97 @msau42

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -77,3 +77,5 @@ replace k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.18.0
 replace k8s.io/sample-cli-plugin => k8s.io/sample-cli-plugin v0.18.0
 
 replace k8s.io/sample-controller => k8s.io/sample-controller v0.18.0
+
+replace github.com/kubernetes-csi/csi-proxy/client => ../../github.com/kubernetes-csi/csi-proxy/client
```
Comment (mauriciopoppe) [inline go.mod:81]: this will be removed once the csi-proxy client is v1

File: manifest_osversion.sh [hunk 2]
```diff
@@ -6,7 +6,6 @@
 # replace the following with annotation approach. https://github.com/docker/cli/pull/2578
 
 export DOCKER_CLI_EXPERIMENTAL=enabled
-_WINDOWS_VERSIONS="20H2 2004 1909 ltsc2019"
```
Comment (mauriciopoppe) [inline manifest_osversion.sh:9]: this line was unused in this program

**Comment by mauriciopoppe** [comment]
/hold

Even though this is working I have to make a small refactor taking all of the methods of both v1 and v1beta mounters to an interface to simplify the code

**Comment by mauriciopoppe** [comment]
/unhold

Made the refactor and run the e2e tests again and added the results in the description

**Comment by mattcary** [comment]
It's the same integration test problem. We still haven't figured out what's going on. After you get an lgtm from Jing I can force-merge this.

File: deploy/kubernetes/overlays/noauth-debug/kustomization.yaml [hunk 3]
```diff
@@ -8,6 +8,7 @@ transformers:
 patchesStrategicMerge:
 - noauth.yaml
 - controller-overlay.yaml
+- node-overlay.yaml
```
Comment (jingxu97) [inline deploy/kubernetes/overlays/noauth-debug/kustomization.yaml:11]: why we need add this yaml too?

File: pkg/gce-pd-csi-driver/utils_windows.go [hunk 4]
```diff
@@ -39,7 +39,7 @@ func formatAndMount(source, target, fstype string, options []string, m *mount.Sa
 // not exist. Currently kubelet creates the path beforehand, this is a workaround to
 // remove the path first.
 func preparePublishPath(path string, m *mount.SafeFormatAndMount) error {
-	proxy, ok := m.Interface.(*mounter.CSIProxyMounter)
+	proxy, ok := m.Interface.(mounter.CSIProxyMounter)
```
Comment (jingxu97) [inline pkg/gce-pd-csi-driver/utils_windows.go:42]: why pointer is removed?

File: deploy/kubernetes/overlays/noauth-debug/kustomization.yaml (hunk 3 above)
Comment (mauriciopoppe) [reply deploy/kubernetes/overlays/noauth-debug/kustomization.yaml:11]: this is a dev only file, I used it to deploy the GCE PD CSI driver with `AlwaysPull`

File: pkg/gce-pd-csi-driver/utils_windows.go (hunk 4 above)
Comment (mauriciopoppe) [reply pkg/gce-pd-csi-driver/utils_windows.go:42]: Previously it was a struct, now it's an interface implemented by v1 and v1beta

**Comment by mattcary** [comment]
I'll wait for the tests to run, confirm they pass except for the usual suspect, and force the merge.

**Pull request merged**